Automattic/pocket-casts-android
# Restrict independent watch functionality to Plus users

## Description
This prevents users from accessing the wear app if they are not Plus users since v1 of the wear app will not have control-the-phone-app-from-the-watch-app functionality for free users.

Getting this working well turned out to be more difficult than I expected. In particular, getting the "logging in" screen to show and block the user until the refresh was complete was the tricky part. I'm mentioning this just to call out that this PR probably merits some extra testing.

A couple of issues you may see:
1. Because the button for proceeding to log in is off screen, I also made the text on that screen clickable to reduce the chance of a user get hung up on that screen.
1. I've noticed that occasionally when installing the app from Android Studio on top of a previous install, then logging out, and then logging back in with google, the app could get stuck on a logging in screen. I've only seen this a couple of times, and I have never reproduced it on a clean install.
3. Because of [this issue](https://github.com/Automattic/pocket-casts-android/issues/958) you may see crashes when logging out (but only on debug builds). For testing purposes, I just [updated the RxJavaUncaughtExceptionHandling object](https://github.com/Automattic/pocket-casts-android/blob/b06c9721e256f74ea31299131929330e16d0c99d/modules/services/utils/src/main/java/au/com/shiftyjelly/pocketcasts/utils/log/RxJavaUncaughtExceptionHandling.kt#L20-L22) to not throw any exceptions because anytime someone logs into the watch app with a free account, they get immediately logged out (which can result in a crash on debug builds). I'm starting to wonder if maybe this [crash-only-in-debug idea of mine](https://github.com/Automattic/pocket-casts-android/pull/971) wasn't such a great idea.
4.  I had it happen a couple of times where the watch app would log in automatically even though the phone app was not logged in. I haven't been able to track down the cause of this (and it doesn't seem like a big problem since the user can just log out if they want to switch accounts).
5. Note that, currently, once you log in using a google account, you [cannot then switch to a different google account](https://github.com/Automattic/pocket-casts-android/issues/993) without reinstalling the app.

## Testing Instructions

For each of the following scenarios:

1. Logging in automatically because your watch-connected phone is logged in
2. Logging in using a Google account
3. Logging in using email and password

Verify that:

1. if it is a free account, the logging in screen will be displayed, and then the user is returned to the UI prompting the user to log in.
2. If it is a plus account, the logging in screen is displayed, and that you are taken into the app once the refresh is complete

In addition, also test that we are handling accounts that change from Plus to Free appropriately:
1. Sign in with a Plus account
2. Change the account to being Free
3. Perform a manual refresh from the Settings screen
4. Verify that you are returned to the `RequirePlusScreen`

### Phone app

Because I made a small change to how we report a user's subscription status, it would be good to do some smoke-testing around logging into and out of some Plus and Free accounts on the phone app.

## Screenshots or Screencast 

| Plus | Free |
| --- | --- |
| <video src="https://github.com/Automattic/pocket-casts-android/assets/4656348/a3710eb7-af98-4119-a0b2-132201476f30"/> | <video src="https://github.com/Automattic/pocket-casts-android/assets/4656348/479d7cab-04d4-4949-accf-af58315137e0"/> |

## Checklist
 
#### I have tested any UI changes...
